packages/publisher/publisher.go

```diff
@@ -11,6 +23,101 @@ var Event = events.NewEvent(func(handler interface{}, params ...interface{}) {
 	callback(msgType, parts)
 })
 
-func Publish(msgType string, parts ...string) {
+////////////////////////////////////////////////////////////////////////////////
+// Publisher
+
+type Publisher struct {
+	blockAppliedPipe pipe.Pipe
+	mutex            *sync.RWMutex
+	running          *atomic.Bool
+	log              *logger.Logger
+}
+
+var _ chain.ChainListener = &Publisher{}
+
+type publisherBlockApplied struct {
+	chainID isc.ChainID
+	block   state.Block
+}
+
+func NewPublisher(log *logger.Logger) *Publisher {
+	p := &Publisher{
+		blockAppliedPipe: pipe.NewDefaultInfinitePipe(),
+		mutex:            &sync.RWMutex{},
+		running:          &atomic.Bool{},
+		log:              log,
+	}
+	return p
+}
+
+// Implements the chain.ChainListener interface.
+// NOTE: Do not block the caller!
+func (p *Publisher) BlockApplied(chainID isc.ChainID, block state.Block) {
+	p.blockAppliedPipe.In() <- &publisherBlockApplied{chainID: chainID, block: block}
+}
+
+// This is called by the component to run this.
+func (p *Publisher) Run(ctx context.Context) {
+	blockAppliedPipeOutCh := p.blockAppliedPipe.Out()
+	ctxDone := ctx.Done()
+	for {
+		select {
+		case blockAppliedUntyped := <-blockAppliedPipeOutCh:
+			p.handleBlockApplied(blockAppliedUntyped.(*publisherBlockApplied))
+		case <-ctxDone:
+			return
+		}
+	}
+}
```

Comment:
minor suggestion (to remove `ctxDone := ctx.Done()`)

```suggestion
func (p *Publisher) Run(ctx context.Context) {
	blockAppliedPipeOutCh := p.blockAppliedPipe.Out()
	for {
		select {
		case blockAppliedUntyped := <-blockAppliedPipeOutCh:
			p.handleBlockApplied(blockAppliedUntyped.(*publisherBlockApplied))
		case <-ctx.Done():
			return
		}
	}
}
```

Reply:
Isn't then the function Done() is executed on each loop iteration?

Reply:
guess so, but that's probably negligible.

that is how the go docs do it:
https://pkg.go.dev/context#Context